src/kaleidoscope/device/dygma/raise/Hand.cpp

```diff
@@ -0,0 +1,322 @@
+/* -*- mode: c++ -*-
+ * kaleidocope::device::dygma::Raise -- Kaleidoscope device plugin for Dygma Raise
+ * Copyright (C) 2017-2019  Keyboard.io, Inc
+ * Copyright (C) 2017-2019  Dygma Lab S.L.
+ *
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifdef ARDUINO_SAMD_RAISE
+
+#include <Arduino.h>
+#include "Hand.h"
+#include <Wire.h>
+#include "crc16.h"
+
+namespace kaleidoscope {
+namespace device {
+namespace dygma {
+namespace raise {
+
+#define SCANNER_I2C_ADDR_BASE 0x58
+#define ELEMENTS(arr)  (sizeof(arr) / sizeof((arr)[0]))
+
+uint8_t twi_uninitialized = 1;
+
+#ifdef __SAMD21G18A__
```

Comment:
this twi implementation ought to get broken out into its own source file, at least.

Reply:
Yep, will clean up the whole file.